man/objdwarf-ng.1
.TH OBJDWARF-NG "1" "October 2026" "objtools 1.10" "User Commands"
.SH NAME
objdwarf-ng: display information about ELF files.

.SH SYNOPSIS
objdwarf-ng [\fB\-H\fR|\fB\-\-help\fR]
            [\fB\-v\fR|\fB\-\-version\fR]
            [\fB\-V\fR|\fB\-\-verbose\fR]
            [\fB\-x\fR <number or name>|\fB\-\-hex\-dump=\fR<number or name>]
            [\fB\-p\fR <number or name>|\fB\-\-string\-dump=\fR<number or name>]
            [\fB\-w\fR[aAcFikKLmrRsptTuUgX]\fR|
            \fB\-\-dwarf\-dump\fR[=abbrev,=addr,=cu_index,=frames\-interp,=info,=links,=follow\-links,
                    =decodedline,=macro,=aranges,=ranges,=str,=pubnames,=pubtypes,=trace_aranges,
                    =trace_abbrev,=trace_info,=gdb_index,=enhanced]]
            \fIelffile\fR...

.SH DESCRIPTION
\fBobjdwarf-ng\fR displays the dwarf information about one or more ELF format object files. The options control what particular information to display.

\fIelffile...\fR are the object files to be examined. 32-bit and 64-bit ELF files are supported, as are archives containing ELF files.

.SH OPTIONS
.IP "\fB\-H\fR" 4
.IX Item "-H"
.PD 0
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD
Print a summary of the options to objdwarf-ng and exit.
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD 0
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD
Print the version number of objdwarf-ng and exit.
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD 0
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD
Print additional information.
.IP "\fB\-x <number or name>\fR" 4
.IX Item "-x <number or name>"
.PD 0
.IP "\fB\-\-hex\-dump=<number or name>\fR" 4
.IX Item "--hex-dump=<number or name>"
.PD
Displays the contents of the indicated section as a hexadecimal bytes.
.IP "\fB\-p <number or name>\fR" 4
.IX Item "-p <number or name>"
.PD 0
.IP "\fB\-\-string\-dump=<number or name>\fR" 4
.IX Item "--string-dump=<number or name>"
.PD
Displays the contents of the indicated section as printable strings.
.IP "\fB\-w[aAcFikKLmrRsptTuUgX]\fR" 4
.IX Item "-w[aAcFikKLmrRsptTuUgX]"
.PD 0
.IP "\fB\-\-dwarf-dump[=abbrev,=addr,=cu_index,=frames-interp,=info,=links,=follow-links,=decodedline,=macro,=aranges,=ranges,=str,=pubnames,=pubtypes,=trace_aranges,=trace_abbrev,=trace_info,=gdb_index,=enhanced]\fR" 4
.IX Item "--dwarf-dump[=abbrev,=addr,=cu_index,=frames-interp,=info,=links,=follow-links,=decodedline,=macro,=aranges,=ranges,=str,=pubnames,=pubtypes,=trace_aranges,=trace_abbrev,=trace_info,=gdb_index,=enhanced]"
.PD
Displays the contents of the \s-1DWARF\s0 debug sections in the file, if any are present.
.RS 4
.ie n .IP """a""" 4
.IX Item "a"
.PD 0
.ie n .IP """=abbrev""" 4
.IX Item "=abbrev"
.PD
Displays the contents of the \fB.debug_abbrev\fR section.
.ie n .IP """A""" 4
.IX Item "A"
.PD 0
.ie n .IP """=addr""" 4
.IX Item "=addr"
.PD
Displays the contents of the \fB.debug_addr\fR section.
.ie n .IP """c""" 4
.IX Item "c"
.PD 0
.ie n .IP """=cu_index""" 4
.IX Item "=cu_index"
.PD
Displays the contents of the \fB.debug_cu_index\fR|\fB.debug_tu_index\fR sections.
.ie n .IP """F""" 4
.IX Item "F"
.PD 0
.ie n .IP """=frames\-interp""" 4
.IX Item "=frames-interp"
.PD
Display the interpreted contents of a \fB.debug_frame\fR section.
.ie n .IP """i""" 4
.IX Item "i"
.PD 0
.ie n .IP """=info""" 4
.IX Item "=info"
.PD
Displays the contents of the \fB.debug_info\fR section.
.ie n .IP """k""" 4
.IX Item "k"
.PD 0
.ie n .IP """=links""" 4
.IX Item "=links"
.PD
Displays the contents of the \fB.gnu_debuglink\fR|\fB.gnu_debugaltlink\fR sections.
.ie n .IP """K""" 4
.IX Item "K"
.PD 0
.ie n .IP """=follow\-links""" 4
.IX Item "=follow-links"
.PD
Display the contents of any selected debug sections that are found in linked, separate debug info file(s).
.ie n .IP """L""" 4
.IX Item "L"
.PD 0
.ie n .IP """=decodedline""" 4
.IX Item "=decodedline"
.PD
Displays the interpreted contents of the \fB.debug_line\fR section.
.ie n .IP """m""" 4
.IX Item "m"
.PD 0
.ie n .IP """=macro""" 4
.IX Item "=macro"
.PD
Displays the contents of the \fB.debug_macro\fR|\fB.debug_macinfo\fR sections.
.ie n .IP """r""" 4
.IX Item "r"
.PD 0
.ie n .IP """=aranges""" 4
.IX Item "=aranges"
.PD
Displays the contents of the \fB.debug_aranges\fR section.
.ie n .IP """R""" 4
.IX Item "R"
.PD 0
.ie n .IP """=ranges""" 4
.IX Item "=ranges"
.PD
Displays the contents of the \fB.debug_ranges\fR|\fB.debug_rnglists\fR sections.
.ie n .IP """s""" 4
.IX Item "s"
.PD 0
.ie n .IP """=str""" 4
.IX Item "=str"
.PD
Displays the contents of the \fB.debug_str\fR|\fB.debug_line_str\fR|\fB.debug_str_offsets\fR sections.
.ie n .IP """p""" 4
.IX Item "p"
.PD 0
.ie n .IP """=pubnames""" 4
.IX Item "=pubnames"
.PD
Displays the contents of the \fB.debug_pubnames\fR|\fB.debug_gnu_pubnames\fR sections.
.ie n .IP """t""" 4
.IX Item "t"
.PD 0
.ie n .IP """=pubtypes""" 4
.IX Item "=pubtypes"
.PD
Displays the contents of the \fB.debug_pubtypes\fR|\fB.debug_gnu_pubtypes\fR sections.
.ie n .IP """T""" 4
.IX Item "T"
.PD 0
.ie n .IP """=trace_aranges""" 4
.IX Item "=trace_aranges"
.PD
Displays the contents of the \fB.trace_aranges\fR section.
.ie n .IP """u""" 4
.IX Item "u"
.PD 0
.ie n .IP """=trace_abbrev""" 4
.IX Item "=trace_abbrev"
.PD
Displays the contents of the \fB.trace_abbrev\fR section.
.ie n .IP """U""" 4
.IX Item "U"
.PD 0
.ie n .IP """=trace_info""" 4
.IX Item "=trace_info"
.PD
Displays the contents of the \fB.trace_info\fR section.
.ie n .IP """g""" 4
.IX Item "g"
.PD 0
.ie n .IP """=gdb_index""" 4
.IX Item "=gdb_index"
.PD
Displays the contents of the \fB.gdb_index\fR|\fB.debug_names\fR sections.
.ie n .IP """X""" 4
.IX Item "X"
.PD 0
.ie n .IP """=enhanced""" 4
.IX Item "=enhanced"
.PD
Displays the contents in the enhanced mode format.

.SH SEE ALSO
detect-ng(1), convert-ng(1), coredump-ng(1), objcopy-ng(1), objdump-ng(1), objdwarf-ng(1), objhash-ng(1), readelf-ng(1), readpe-ng(1)

.SH COPYRIGHT
.IX Header "COPYRIGHT"
MIT License
.PP
Copyright (c) 2025 Kyle Wickens

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.



CHANGELOG
v1.2
-----------------------------------------------------------------------
[+] Added big-endian support.
[+] Added MIPS architecture support.
[+] Improved RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.1
-----------------------------------------------------------------------
[+] Added ARM Thumb architecture support.
[+] Added RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.0
-----------------------------------------------------------------------
[+] Added ARM architecture support.
[+] Improved x86-64 architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################



src/objdwarf-ng.c
#include "buffer.h"
#include "printf.h"
#include "readar.h"
#include "readelf.h"
#include "options.h"

#include "static/objdwarf-ng.ci"

#define THIS_NAME "objdwarf-ng"

static int usage(poptions_t o, const char* name, const args_t args[], const int ecode) {
  int n = 0;
  n += usage_name(o, name, args, zDESCRIPTIONELF);
  n += usage_synopsis0(o, name, args);
  n += usage_synopsis1(o, name, zDEBUGELFARGS, zOBJDWARFARGS0, zOBJDWARFARGS1);
  n += usage_objname(o, name, "elffile");
  n += printf_eol();
  n += usage_description(o, name, args, zOBJDWARFDESCRIPTION);
  n += usage_options0(o, name, args);
  n += usage_options1(o, name, zDEBUGELFARGS, zOBJDWARFARGS0, zOBJDWARFARGS1);
  n += usage_seealso(o, name, args);
  n += usage_copyright(o, name, args);

  return ecode;
}

static int get_options_objdwarf(poptions_t o, int argc, char** argv, char* name) {
  if (0 == argc) {
    return usage(o, THIS_NAME, zOBJDWARFARGS, ECODE_ARGUMENTS);
  }

  strname(o->prgname, name);
  printf_errname(o->prgname);

  for (int i = 0; i < argc; ++i) {
    if ('-' == argv[i][0] && '-' == argv[i][1]) {
      MALLOCA(char, arg0, 1024);
      MALLOCA(char, arg1, 1024);

      if (ECODE_ISOK(breakup_args(argv[i], arg0, NELEMENTS(arg0), arg1, NELEMENTS(arg1)))) {
        if (0 == xstrcmp(arg0, zOBJDWARFARGS1)) {
          imode_t ocdump = get_options2(o, zDEBUGELFARGS, arg1);
          if (0 == ocdump) {
            return odeath(o, THIS_NAME, arg1);
          }
          o->ocdump |= ocdump;
        } else if (0 == xstrcmp(arg0, "--hex-dump")) {
          oinsertsname(o, ACT_HEXDUMP, arg1);
        } else if (0 == xstrcmp(arg0, "--string-dump")) {
          oinsertsname(o, ACT_STRDUMP8, arg1);
        } else if (0 == xstrcmp(arg0, "--decompress")) {
          oinsertsname(o, ACT_ZLIB, arg1);
        } else {
          return odeath(o, THIS_NAME, arg0 + 2);
        }
      } else {
        imode_t action = get_options2(o, zOBJDWARFARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 2);
        }
        o->action |= action;
      }
    } else if ('-' == argv[i][0]) {
      if (argv[i][0] == zOBJDWARFARGS0[0] && argv[i][1] == zOBJDWARFARGS0[1]) {
        imode_t ocdump = get_options1(o, zDEBUGELFARGS, argv[i] + 1);
        if (0 == ocdump && argv[i][2]) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        o->ocdump |= ocdump ? ocdump : set_options1(o, zDEBUGELFARGS);
      } else if (0 == xstrcmp(argv[i], "-x")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_HEXDUMP, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-p")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_STRDUMP8, argv[++i]);
      } else if (0 == xstrcmp(argv[i], "-z")) {
        if (argc <= (i + 1)) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        oinsertsname(o, ACT_ZLIB, argv[++i]);
      } else {
        imode_t action = get_options1(o, zOBJDWARFARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        o->action |= action;
      }
    } else if (0 == o->inpname[0]) {
      xstrncpy(o->inpname, argv[i], NELEMENTS(o->inpname));
    } else {
      return odeath(o, THIS_NAME, argv[i]);
    }
  }

  if (o->action & OPTPROGRAM_VERSION) {
    return version0(o, THIS_NAME, zOBJDWARFARGS);
  }

  if (o->action & OPTPROGRAM_HELP) {
    return usage(o, THIS_NAME, zOBJDWARFARGS, ECODE_OK);
  }

  return ECODE_OK;
}

int main(int argc, char* argv[]) {
  int x = ECODE_MALLOC;
  poptions_t o = omalloc();
  if (o) {
    x = get_options_objdwarf(o, argc - 1, argv + 1, argv[0]);
    if (ECODE_ISOK(x) && o->inpname[0]) {
      pbuffer_t p = bopen(o->inpname);
      if (p) {
        if (isAR(p)) {
          x = readar(p, o);
        } else if (isELF(p)) {
          x = readelf(p, o);
        } else {
          printf_e("'%s': invalid file format.", o->inpname);
        }
      } else {
        printf_e("'%s': no such file.", o->inpname);
      }

      bfree(p);
    }

    ofree(o);
  }

  return x;
}

